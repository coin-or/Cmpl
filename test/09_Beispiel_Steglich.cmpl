//
// Produktionsprogrammplanung unter Einbeziehung von Lernkurveneffekten
//
// Beispiel von Mike Steglich 20.10.2007
//

const:
	// Beispielhaft wird hier von 2 Produkten, 3 Produktionsfaktoren und 2 Arbeitsplätzen ausgegangen
	Produkte = 2;
	Faktoren = 3;
	Plaetze = 2;

	// Absatzpreis
 	Absatzpreis[1..Produkte] = (20, 30);
 
 	// Stückkosten ohne Fertigungseinzelkosten
 	Stueckkosten[1..Produkte] = (10, 18);
 
 	// Kostensatz pro Zeiteinheit für die Arbeit pro Arbeitsplatz
 	Kostensatz[1..Plaetze] = (1.3, 1.5);
 
 	// Produktionskoeffizienten pro Produktionsfaktor und Produkt
 	A[1..Faktoren, 1..Produkte] = ((1.0, 1.0), (1.2, 1.8), (1.4, 1.3));
 
 	// Obergrenze pro Produktionsfaktor
 	B[1..Faktoren] = (40, 60, 70);
 
 	// Maximale Personalkapazität pro Arbeitsplatz
 	T[1..Plaetze] = (80, 100);
 
 	// Maximale Menge pro Produkt
 	xMax[1..Produkte] = (100, 100);
 
 	// Produktionsmenge pro Linearisierungsintervall
 	// Für das Produkt 1 gibt es 2 Linearisierungsintervalle, für das Produkt 2 gibt es 3
 	xniMax[1..Produkte, ] = ((3, 100), (2, 4, 100));
 	
 	// Produktionskoeffizienten pro Arbeitsplatz, Produkt und Linearisierungsintervall
 	// (Die Werte müssen direkt hier eingetragen werden, weil die log-Funktion zur Berechnung
 	// aus Ausgangswerten hier nicht implementiert ist (die übrige Berechnung wäre hier möglich))
 	tkni[1..Plaetze, 1..Produkte, ] =
 		(((1.6, 1.2), (2.1, 1.8, 1.7)),
 		 ((2.5, 2.2), (3.2, 2.8, 2.6)));

echo "QQQ";

include "a/test.gen"

echo "XXX";

include "09_Include.gen"

 var:
 	// Produktionsprogramm
 	x[1..Produkte] : real[0..];
 
 	// Gesamte zeitliche Beanspruchung der Arbeit pro Arbeitsplatz und Produkt
 	rkn[1..Plaetze, 1..Produkte] : real[0..];
 
 	// (Hier Schleife über Produkte, da die Produkte unterschiedlich viele Linearisierungsintervalle haben)
 	{n = 1(1)Produkte:
 		// Produktion pro Linearisierungsstück
 		xni[n, 1..dim(xniMax[n,])] : real[0..];
 
 		// Hilfsvariablen für Reihenfolge der Linearisierungsintervalle
 		// (Jeweils die letzte y-Variable taucht in den Restriktionen gar nicht auf,
 		//  sie wird deshalb trotz der Definition hier gar nicht in die generierte Matrix aufgenommen)
 		y[n, 1..dim(xniMax[n,])] : binary;
 	}
 
 
 // Zielfunktion
 objective:
 	sum{n=1(1)Produkte: (Absatzpreis[n] - Stueckkosten[n]) * x[n]}
 		+ sum{n=1(1)Produkte, k=1(1)Plaetze: -Kostensatz[k] * rkn[k,n]};
 	
 // Restriktionen
 constraint:
 	// Basisproblem über Produktionsfaktoren
 	A[,] * x[] <= B[];
 
 	// Berechnung der rkn
 	RS {k=1(1)Plaetze, n=1(1)Produkte:
 		//RS$k$x$n$:
			rkn[k,n] = sum{i=1(1)dim(xniMax[n,]): tkni[k,n,i] * xni[n,i]};
 	}
 
 	// Beschränkung der verfügbaren Zeitkapazität
 	{k=1(1)Plaetze:
 		sum{n=1(1)Produkte: rkn[k,n]} <= T[k];
 	}
 
 	// Bedingung für Reihenfolge der Linearisierungsintervalle
 	{n=1(1)Produkte:
 		x[n] = sum{i=1(1)dim(xniMax[n,]): xni[n,i]};
 
 		{i=1(1)dim(xniMax[n,]):
 			{i < dim(xniMax[n,]): xni[n,i] >= y[n,i] * xniMax[n,i]; }
 			{i > 1:               xni[n,i] <= y[n,i-1] * xniMax[n,i]; }
 		}
 	}
 
 	// Schranken
 	0 <= x[] <= xMax[];
 	{n=1(1)Produkte: 0 <= xni[n,] <= xniMax[n,]; }
	
