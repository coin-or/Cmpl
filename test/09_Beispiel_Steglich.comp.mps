* CMPL - Free-MPS - Export
NAME          09_Beispiel_Steglich.cmpl
* OBJNAME line_1
* OBJSENSE max
ROWS
 N line_1
 L line_2
 L line_3
 L line_4
 E RS_1_1
 E RS_1_2
 E RS_2_1
 E RS_2_2
 L line_9
 L line_10
 E line_11
 G line_12
 L line_13
 E line_14
 G line_15
 G line_16
 L line_17
 L line_18
COLUMNS
 x[1] line_1 10 line_2 1.000000
 x[1] line_3 1.200000 line_4 1.400000
 x[1] line_11 1
 x[2] line_1 12 line_2 1.000000
 x[2] line_3 1.800000 line_4 1.300000
 x[2] line_14 1
 rkn[1,1] line_1 -1.300000 RS_1_1 1
 rkn[1,1] line_9 1
 rkn[1,2] line_1 -1.300000 RS_1_2 1
 rkn[1,2] line_9 1
 rkn[2,1] line_1 -1.500000 RS_2_1 1
 rkn[2,1] line_10 1
 rkn[2,2] line_1 -1.500000 RS_2_2 1
 rkn[2,2] line_10 1
 xni[1,1] RS_1_1 -1.600000 RS_2_1 -2.500000
 xni[1,1] line_11 -1 line_12 1
 xni[1,2] RS_1_1 -1.200000 RS_2_1 -2.200000
 xni[1,2] line_11 -1 line_13 1
 xni[2,1] RS_1_2 -2.100000 RS_2_2 -3.200000
 xni[2,1] line_14 -1 line_15 1
 xni[2,2] RS_1_2 -1.800000 RS_2_2 -2.800000
 xni[2,2] line_14 -1 line_16 1
 xni[2,2] line_17 1
 xni[2,3] RS_1_2 -1.700000 RS_2_2 -2.600000
 xni[2,3] line_14 -1 line_18 1
 GVANF 'MARKER' 'INTORG'
 y[1,1] line_12 -3 line_13 -100
 y[2,1] line_15 -2 line_17 -4
 y[2,2] line_16 -4 line_18 -100
 GVEND 'MARKER' 'INTEND'
RHS
 RHS line_2 40 line_3 60
 RHS line_4 70 line_9 80
 RHS line_10 100
RANGES
BOUNDS
 UP BOUND x[1] 100
 UP BOUND x[2] 100
 PL BOUND rkn[1,1]
 PL BOUND rkn[1,2]
 PL BOUND rkn[2,1]
 PL BOUND rkn[2,2]
 UP BOUND xni[1,1] 3
 UP BOUND xni[1,2] 100
 UP BOUND y[1,1] 1
 UP BOUND xni[2,1] 2
 UP BOUND xni[2,2] 4
 UP BOUND xni[2,3] 100
 UP BOUND y[2,1] 1
 UP BOUND y[2,2] 1
ENDATA
