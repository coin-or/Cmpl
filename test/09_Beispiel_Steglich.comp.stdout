QQQ
Testausgabe1
qqq
Testausgabe2
XXX
Test3 Test
